File: src/main/java/ezgames/immatcher/matchers/BasicMatchers.java
package ezgames.immatcher.matchers;

import static ezgames.immatcher.Matchers.*;
import ezgames.immatcher.Matcher;

/**
 * {@code BasicMatchers} is a collection of static methods for creating
 * {@code Matcher}s that test basic things that apply to all objects or primitives.
 */
public class BasicMatchers {
	/**
	 * Returns a {@code Matcher} that tests that the object under test is null.
	 * @return a {@code Matcher} that tests for null
	 */
	public static <T> Matcher<T> isNull() {
		return IsNull.isNull();
	}
	
	/**
	 * Returns a {@code Matcher} that tests that the object under test is not null.
	 * @return a {@code Matcher} that tests for non-null
	 */
	public static <T> Matcher<T> isNotNull() {
		return not(isNull());
	}
	
	/**
	 * Returns a {@code Matcher} that tests that the object under test is equal to
	 * the given object.
	 * @param object - object to test for equality against
	 * @return a {@code Matcher} that tests for equality
	 */
	public static <T> Matcher<T> isEqualTo(T object) {
		return Equals.isEqualTo(object);
	}
	
	/**
	 * Returns a {@code Matcher} that tests that the object under test is not
	 * equal to the given object.
	 * @param object - object to test for non-equality against
	 * @return a {@code Matcher} that tests for non-equality
	 */
	public static <T> Matcher<T> isNotEqualTo(T object) {
		return invert(isEqualTo(object), "wasn't equal to " + object.toString());
	}
}

File: src/main/java/ezgames/immatcher/matchers/BasicMatchers.kt
@file:JvmName("BasicMatchers")

package ezgames.immatcher.matchers

import ezgames.immatcher.*
import ezgames.immatcher.Matcher
import ezgames.immatcher.not
import ezgames.immatcher.Result
import ezgames.immatcher.ResultBuilder

class IsNull<in T> internal constructor() : Matcher<T?> {
    override fun match(actual: T?): Result {
        val result = ResultBuilder.withMessages("was Null", "wasn't Null")
        if (actual == null)
            return result.pass()
        else
            return result.fail()
    }

    override fun notMatches(actual: T?): Result {
        val result = ResultBuilder.withMessages("wasn't Null", "was Null")
        if (actual == null)
            return result.fail()
        else
            return result.pass()
    }
}

class Equals<in T> internal constructor(private val obj: T) : Matcher<T> {

    private val expected: String

    init {
        this.expected = "equaled $obj"
    }

    override fun match(actual: T): Result {
        val result = ResultBuilder.withMessages(expected, "equaled $actual")
        if (obj == actual)
            return result.pass()
        else
            return result.fail()
    }

    override fun notMatches(actual: T): Result {
        val result = ResultBuilder.withMessages("equaled $actual", expected)
        if (obj == actual)
            return result.fail()
        else
            return result.pass()
    }
}

/**
 * Returns a `Matcher` that tests that the object under test is null.
 * @return a `Matcher` that tests for null
 */
fun <T> isNull(): Matcher<T> {
    return IsNull()
}

/**
 * Returns a `Matcher` that tests that the object under test is not null.
 * @return a `Matcher` that tests for non-null
 */
fun <T> isNotNull(): Matcher<T> {
    return not(isNull<T>())
}

/**
 * Returns a `Matcher` that tests that the object under test is equal to
 * the given object.
 * @param object - object to test for equality against
 * *
 * @return a `Matcher` that tests for equality
 */
fun <T> isEqualTo(expected: T): Matcher<T> {
    return Equals(expected)
}

/**
 * Returns a `Matcher` that tests that the object under test is not
 * equal to the given object.
 * @param object - object to test for non-equality against
 * *
 * @return a `Matcher` that tests for non-equality
 */
fun <T> isNotEqualTo(expected: T): Matcher<T> {
    return not(Equals(expected))
}
